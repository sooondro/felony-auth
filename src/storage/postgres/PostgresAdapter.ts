import { Sequelize, DataTypes } from 'sequelize'
import Bcrypt from 'bcrypt'

import Authentication from '../../Authentication'
import Models from './db/models/Models'
import { ValidationErrors } from '../../error/ValidationError'

import StorageAdapterInterface from '../StorageAdapterInterface'
import RegistrationData from '../../types/RegistrationData'
import LoginData from '../../types/LoginData'
import PostgresConnectionData from '../../types/PostgresConnectionData'
import TwoFactorRegistrationData from '../../types/TwoFactorRegistrationData'
import AuthenticableUser from '../../types/AuthenticableUser'
import AuthenticableTwoFactorUser from '../../types/AuthenticableTwoFactorUser'
import AuthenticationError from '../../error/AuthenticationError'

/**
 * Storage adapter for the Postgres database.
 */
export default class PostgresAdapter implements StorageAdapterInterface {
  private client!: Sequelize
  private authentication!: Authentication
  public models!: Models

  /**
   * Getter method for Postgres client.
   */
  public get Client (): Sequelize {
    return this.client
  }

  /**
   * Used for injecting Authentication class into the adapter.
   *
   * @param {Authentication} authentication
   */
  initialize (authentication: Authentication): void {
    this.authentication = authentication
  }

  /**
   * Set up Postgres connection with the config object.
   *
   * @param {PostgresConnectionData} config
   * @throws
   */
  async setupConnectionWithConnectionData (config: PostgresConnectionData): Promise<void> {
    this.client = new Sequelize(config.database, config.username, config.password, {
      dialect: 'postgres',
      host: config.host,
      port: config.port,
      logging: false
    })

    this.models = new Models(this.client, DataTypes)
  }

  /**
   * Set up Postgres connection with the config object.
   *
   * @param {string} connectionUrl
   * @throws
   */
  async setupConnectionWithConnectionUrl (connectionUrl: string): Promise<void> {
    this.client = new Sequelize(connectionUrl, {
      dialect: 'postgres',
      logging: false
    })

    this.models = new Models(this.client, DataTypes)
  }

  /**
   * Add new user to the database.
   *
   * @param {RegistrationData} payload
   * @return {Promise<AuthenticableUser>}
   * @throws
   */
  async register (payload: RegistrationData): Promise<AuthenticableUser> {
    const hashedPassword = await Bcrypt.hash(payload.password, 12)

    const [user, created]: [AuthenticableUser, boolean] = await this.models.User.findOrCreate({
      where: { email: payload.email },
      defaults: {
        username: payload.username,
        firstName: payload.firstName,
        lastName: payload.lastName,
        email: payload.email,
        password: hashedPassword
      }
    })

    if (!created) {
      const validationErrors = new ValidationErrors()
      validationErrors.addError('email', 'invalid credentials')
      throw validationErrors
    }

    return user
  }

  /**
   * Register two-factor user.
   *
   * @param {TwoFactorRegistrationData} payload
   * @return {Promise<AuthenticableTwoFactorUser>}
   * @throws
   */
  async registerTwoFactorUser (payload: TwoFactorRegistrationData): Promise<AuthenticableTwoFactorUser> {
    const [user, created]: [AuthenticableTwoFactorUser, boolean] = await this.models.TwoFactorUser.findOrCreate({
      where: { userId: payload.userId, provider: payload.provider },
      defaults: {
        userId: payload.userId,
        provider: payload.provider,
        secret: payload.secret
      }
    })

    if (!created) {
      const validationErrors = new ValidationErrors()
      validationErrors.addError('provider', 'user with the given provider already exists')
      throw validationErrors
    }

    return user
  }

  /**
   * Login user.
   *
   * @param {LoginData} payload
   * @return {Promise<AuthenticableUser>}
   * @throws
   */
  async login (payload: LoginData): Promise<AuthenticableUser> {
    const user = await this.models.User.findOne({
      where: { email: payload.email },
      include: {
        model: this.models.TwoFactorUser
      }
    })

    if (user === null) {
      const validationErrors = new ValidationErrors()
      validationErrors.addError('email', 'invalid credentials')
      throw validationErrors
    }

    const result = await Bcrypt.compare(payload.password, user.password)

    if (!result) {
      throw new AuthenticationError('invalid credentials', { name: 'AuthenticationError', statusCode: 401 })
    }

    const authUser: AuthenticableUser = user

    return authUser
  }

  /**
   * Fetch user from the database by email.
   *
   * @param {string} email
   * @return {Promise<AuthenticableUser>}
   * @throws
   */
  async getUserByEmail (email: string): Promise<AuthenticableUser> {
    const user = await this.models.User.findOne({
      where: { email }
    })

    if (user === null) {
      const validationErrors = new ValidationErrors()
      validationErrors.addError('email', 'invalid credentials')
      throw validationErrors
    }

    const authUser: AuthenticableUser = user

    return authUser
  }

  /**
   * Fetch user from the database by id.
   *
   * @param {string} id
   * @return {Promise<AuthenticableUser>}
   * @throws Login Error
   */
  async getUserById (id: string): Promise<AuthenticableUser> {
    const user = await this.models.User.findByPk(id)

    if (user === null) {
      const validationErrors = new ValidationErrors()
      validationErrors.addError('id', 'invalid credentials')
      throw validationErrors
    }

    const authUser: AuthenticableUser = user

    return authUser
  }

  /**
   * Fetch user from the database by username.
   *
   * @param {string} username
   * @return {Promise<AuthenticableUser>}
   */
  async getUserByUsername (username: string): Promise<AuthenticableUser> {
    const user = await this.models.User.findOne({
      where: { username }
    })

    if (user === null) {
      const validationErrors = new ValidationErrors()
      validationErrors.addError('username', 'invalid credentials')
      throw validationErrors
    }

    const authUser: AuthenticableUser = user

    return authUser
  }

  /**
   * Fetch two-factor user by AuthenticableUser object.
   *
   * @param {AuthenticableUser} user
   * @return {AuthenticableTwoFactorUser}
   */
  async getTwoFactorUser (user: AuthenticableUser): Promise<AuthenticableTwoFactorUser> {
    const twoFactorUser: AuthenticableTwoFactorUser = await this.models.TwoFactorUser.findOne({
      where: { userId: user.id }
    })

    if (twoFactorUser === null) {
      const validationErrors = new ValidationErrors()
      validationErrors.addError('id', 'invalid credentials')
      throw validationErrors
    }

    return twoFactorUser
  }

  // /**
  //  * Fetch two-factor user from the database by email.
  //  *
  //  * @param {string} email
  //  * @return {Promise<AuthenticableTwoFactorUser>}
  //  */
  // async getTwoFactorUserByEmail(email: string): Promise<AuthenticableTwoFactorUser> {
  //   const user = await this.models.TwoFactorUser.findOne({
  //     where: { email: email }
  //   });

  //   if (!user) {
  //     const validationErrors = new ValidationErrors();
  //     validationErrors.addError("email", "invalid credentials");
  //     throw validationErrors;
  //   }

  //   const authUser: AuthenticableTwoFactorUser = user;

  //   return authUser;
  // }

  /**
   * Change user's password.
   *
   * @param {string} email
   * @param {string} oldPassword
   * @param {string} newPassword
   * @return {Promise<void>}
   */
  async changePassword (email: string, oldPassword: string, newPassword: string): Promise<void> {
    const user = await this.models.User.findOne({
      where: { email }
    })

    if (user === null) {
      const validationErrors = new ValidationErrors()
      validationErrors.addError('email', 'invalid credentials')
      throw validationErrors
    }

    const result = await Bcrypt.compare(oldPassword, user.password)

    if (!result) {
      throw new AuthenticationError('invalid credentials', { name: 'AuthenticationError', statusCode: 401 })
    }

    const newHashedPassword = await Bcrypt.hash(newPassword, 12)

    await this.models.User.update(
      { password: newHashedPassword },
      { where: { email } }
    )
  }
}
