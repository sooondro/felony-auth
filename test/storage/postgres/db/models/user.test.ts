import { Sequelize, ValidationError } from "sequelize";
import PostgresAdapter from "../../../../../src/storage/postgres/PostgresAdapter";
import mockdate from 'mockdate';

describe("user", () => {
  const postgresAdapter = new PostgresAdapter();

  beforeEach(async () => {
    await postgresAdapter
      .setupPostgresConnectionWithConnectionUri("postgres://postgres:postgrespw@127.0.0.1:5432/felony_auth_test");
    await postgresAdapter.models.TwoFactorUser.destroy({ where: {} });
    await postgresAdapter.models.User.destroy({ where: {} });
  });

  it("should throw errors when no data is provided", async () => {
    // used to ensure consistency of snapshots between test runs
    mockdate.set(0);

    try {
      const user = await postgresAdapter.models.User.create({
        id: "foo"
      });
    } catch (error) {
      if(error instanceof ValidationError) {
        expect(error.errors).toMatchSnapshot();
      }

      mockdate.reset();
    }
  });

  it("should throw errors when empty strings are provided", async () => {
    // used to ensure consistency of snapshots between test runs
    mockdate.set(0);
    try {
      const user = await postgresAdapter.models.User.create({
        id: "",
        username: "",
        firstName: "",
        lastName: "",
        email: "",
        password: "",
      });
    } catch (error) {
      if(error instanceof ValidationError) {
        expect(error.errors).toMatchSnapshot();
      }
    
      mockdate.reset();
    }
  });
});