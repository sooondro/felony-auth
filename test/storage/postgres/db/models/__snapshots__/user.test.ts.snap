// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`user should throw errors when empty strings are provided 1`] = `
Array [
  ValidationErrorItem {
    "instance": Object {
      "createdAt": 1970-01-01T00:00:00.000Z,
      "email": "",
      "firstName": "",
      "id": "",
      "lastName": "",
      "password": "",
      "updatedAt": 1970-01-01T00:00:00.000Z,
      "username": "",
    },
    "message": "Validation isUUID on id failed",
    "origin": "FUNCTION",
    "original": [Error: Validation isUUID on id failed],
    "path": "id",
    "type": "Validation error",
    "validatorArgs": Array [
      4,
    ],
    "validatorKey": "isUUID",
    "validatorName": "isUUID",
    "value": "",
  },
  ValidationErrorItem {
    "instance": Object {
      "createdAt": 1970-01-01T00:00:00.000Z,
      "email": "",
      "firstName": "",
      "id": "",
      "lastName": "",
      "password": "",
      "updatedAt": 1970-01-01T00:00:00.000Z,
      "username": "",
    },
    "message": "Validation notEmpty on username failed",
    "origin": "FUNCTION",
    "original": [Error: Validation notEmpty on username failed],
    "path": "username",
    "type": "Validation error",
    "validatorArgs": Array [
      true,
    ],
    "validatorKey": "notEmpty",
    "validatorName": "notEmpty",
    "value": "",
  },
  ValidationErrorItem {
    "instance": Object {
      "createdAt": 1970-01-01T00:00:00.000Z,
      "email": "",
      "firstName": "",
      "id": "",
      "lastName": "",
      "password": "",
      "updatedAt": 1970-01-01T00:00:00.000Z,
      "username": "",
    },
    "message": "Validation notEmpty on firstName failed",
    "origin": "FUNCTION",
    "original": [Error: Validation notEmpty on firstName failed],
    "path": "firstName",
    "type": "Validation error",
    "validatorArgs": Array [
      true,
    ],
    "validatorKey": "notEmpty",
    "validatorName": "notEmpty",
    "value": "",
  },
  ValidationErrorItem {
    "instance": Object {
      "createdAt": 1970-01-01T00:00:00.000Z,
      "email": "",
      "firstName": "",
      "id": "",
      "lastName": "",
      "password": "",
      "updatedAt": 1970-01-01T00:00:00.000Z,
      "username": "",
    },
    "message": "Validation notEmpty on lastName failed",
    "origin": "FUNCTION",
    "original": [Error: Validation notEmpty on lastName failed],
    "path": "lastName",
    "type": "Validation error",
    "validatorArgs": Array [
      true,
    ],
    "validatorKey": "notEmpty",
    "validatorName": "notEmpty",
    "value": "",
  },
  ValidationErrorItem {
    "instance": Object {
      "createdAt": 1970-01-01T00:00:00.000Z,
      "email": "",
      "firstName": "",
      "id": "",
      "lastName": "",
      "password": "",
      "updatedAt": 1970-01-01T00:00:00.000Z,
      "username": "",
    },
    "message": "Validation isEmail on email failed",
    "origin": "FUNCTION",
    "original": [Error: Validation isEmail on email failed],
    "path": "email",
    "type": "Validation error",
    "validatorArgs": Array [
      Object {
        "allow_display_name": false,
        "allow_utf8_local_part": true,
        "blacklisted_chars": "",
        "host_blacklist": Array [],
        "ignore_max_length": false,
        "require_display_name": false,
        "require_tld": true,
      },
    ],
    "validatorKey": "isEmail",
    "validatorName": "isEmail",
    "value": "",
  },
  ValidationErrorItem {
    "instance": Object {
      "createdAt": 1970-01-01T00:00:00.000Z,
      "email": "",
      "firstName": "",
      "id": "",
      "lastName": "",
      "password": "",
      "updatedAt": 1970-01-01T00:00:00.000Z,
      "username": "",
    },
    "message": "Validation notEmpty on email failed",
    "origin": "FUNCTION",
    "original": [Error: Validation notEmpty on email failed],
    "path": "email",
    "type": "Validation error",
    "validatorArgs": Array [
      true,
    ],
    "validatorKey": "notEmpty",
    "validatorName": "notEmpty",
    "value": "",
  },
  ValidationErrorItem {
    "instance": Object {
      "createdAt": 1970-01-01T00:00:00.000Z,
      "email": "",
      "firstName": "",
      "id": "",
      "lastName": "",
      "password": "",
      "updatedAt": 1970-01-01T00:00:00.000Z,
      "username": "",
    },
    "message": "Validation notEmpty on password failed",
    "origin": "FUNCTION",
    "original": [Error: Validation notEmpty on password failed],
    "path": "password",
    "type": "Validation error",
    "validatorArgs": Array [
      true,
    ],
    "validatorKey": "notEmpty",
    "validatorName": "notEmpty",
    "value": "",
  },
  ValidationErrorItem {
    "instance": Object {
      "createdAt": 1970-01-01T00:00:00.000Z,
      "email": "",
      "firstName": "",
      "id": "",
      "lastName": "",
      "password": "",
      "updatedAt": 1970-01-01T00:00:00.000Z,
      "username": "",
    },
    "message": "Validation len on password failed",
    "origin": "FUNCTION",
    "original": [Error: Validation len on password failed],
    "path": "password",
    "type": "Validation error",
    "validatorArgs": Array [
      6,
      256,
    ],
    "validatorKey": "len",
    "validatorName": "len",
    "value": "",
  },
]
`;

exports[`user should throw errors when no data is provided 1`] = `
Array [
  ValidationErrorItem {
    "instance": Object {
      "createdAt": 1970-01-01T00:00:00.000Z,
      "id": "foo",
      "updatedAt": 1970-01-01T00:00:00.000Z,
    },
    "message": "User.username cannot be null",
    "origin": "CORE",
    "path": "username",
    "type": "notNull Violation",
    "validatorArgs": Array [],
    "validatorKey": "is_null",
    "validatorName": null,
    "value": null,
  },
  ValidationErrorItem {
    "instance": Object {
      "createdAt": 1970-01-01T00:00:00.000Z,
      "id": "foo",
      "updatedAt": 1970-01-01T00:00:00.000Z,
    },
    "message": "User.email cannot be null",
    "origin": "CORE",
    "path": "email",
    "type": "notNull Violation",
    "validatorArgs": Array [],
    "validatorKey": "is_null",
    "validatorName": null,
    "value": null,
  },
  ValidationErrorItem {
    "instance": Object {
      "createdAt": 1970-01-01T00:00:00.000Z,
      "id": "foo",
      "updatedAt": 1970-01-01T00:00:00.000Z,
    },
    "message": "User.password cannot be null",
    "origin": "CORE",
    "path": "password",
    "type": "notNull Violation",
    "validatorArgs": Array [],
    "validatorKey": "is_null",
    "validatorName": null,
    "value": null,
  },
  ValidationErrorItem {
    "instance": Object {
      "createdAt": 1970-01-01T00:00:00.000Z,
      "id": "foo",
      "updatedAt": 1970-01-01T00:00:00.000Z,
    },
    "message": "Validation isUUID on id failed",
    "origin": "FUNCTION",
    "original": [Error: Validation isUUID on id failed],
    "path": "id",
    "type": "Validation error",
    "validatorArgs": Array [
      4,
    ],
    "validatorKey": "isUUID",
    "validatorName": "isUUID",
    "value": "foo",
  },
]
`;
